### PR TITLE
(#4) add helper for enforcing required properties

### PR DESCRIPTION
**Pull request by eli-darkly**
The need to write code along the lines of "we've read all the properties of the object - now check whether required property X was found or not" turned out to be common enough, as I was implementing unmarshalers in the SDK, that this helper seemed like a desirable feature.

### CODE REVIEW

File: jreader/reader_benchmark_test.go [hunk 1]
```diff
@@ -144,6 +144,18 @@ func BenchmarkReadObjectNoAlloc(b *testing.B) {
 	}
 }
 
+func BenchmarkReadObjectWithRequiredPropsNoAlloc(b *testing.B) {
```
Comment (eli-darkly) [inline jreader/reader_benchmark_test.go:172]: There's a rule in the makefile that causes an error if any benchmark with a suffix of "NoAlloc" ends up causing a heap allocation. Using this new required properties feature shouldn't incur any heap overhead unless you specify more than 20 required properties (that being the size of the preallocated boolean array in ObjectState).

File: jreader/reader.go [hunk 2]
```diff
@@ -48,6 +48,15 @@ func (r *Reader) AddError(err error) {
 	}
 }
 
+// ReplaceError sets the Reader's error value and puts it into a failed state, replacing any
+// previously reported error. If the parameter is nil, it does nothing (a failed state cannot be
+// changed to a non-failed state).
+func (r *Reader) ReplaceError(err error) {
```
Comment (eli-darkly) [inline jreader/reader.go:54]: This method is necessary in order for unmarshalers to be able to change a generic "property X is missing" error to some other error type that they prefer. We do this in go-sdk-common where there's a special error for a user not having a key.

File: internal/commontest/common_test_data.go [hunk 3]
```diff
@@ -66,3 +67,20 @@ func MakeStringsJSON(strings []string) []byte {
 	buf.WriteRune(']')
 	return buf.Bytes()
 }
+
```
Comment (eli-darkly) [inline internal/commontest/common_test_data.go:70]: This is for a new benchmark that I added below.

File: jreader/json_unmarshal_comparative_benchmark_test.go [hunk 4]
```diff
@@ -146,3 +147,25 @@ func benchmarkReadObjectJSONUnmarshal(b *testing.B) {
 		}
 	}
 }
+
+func benchmarkReadArrayOfObjectsJSONUnmarshal(b *testing.B) {
```
Comment (eli-darkly) [inline jreader/json_unmarshal_comparative_benchmark_test.go:151]: This is tangential to the current PR, although it does relate to the API for reading JSON objects. I'm just adding a benchmark for parsing a big list of structs, as a slightly better simulation of the larger data structures that the SDK is likely to parse; the existing benchmark of a single struct is so small that it's harder to compare.

**Review by hroederld — APPROVED** [review]

**Pull request merged**